sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -143,23 +159,44 @@ public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
                 return _segments.GetEnumerator();
             }
 
-            private void Append(Data segment)
+            private void Append(IEnumerable<Data> dataSegments)
             {
+                int length = 0;
+                int numberOfSegments = 0;
+                List<ReadOnlyMemory<byte>> segments = null;
```

Comment:
Just to make sure that I'm understanding - the main changes are here, calling once rather than multiple times to reduce allocations and right-sizing the set of segments?

Reply:
The main changes are:

- Allocating the segments list with the right number of items because we mostly know the data segments upfront
- Precalculating the required length upfront
- Instantiating the `ArrayBufferWriter<byte>`  with the known required length to avoid it growing during the calls to `GetMemory` and saving additional allocations because the `CheckAndResizeBuffer` potentially allocates more than we ever need due to [doubling the array size](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L176-L177)